nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2022 Joshua Wallace
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <utility>
+
+#include "assisted_teleop.hpp"
+#include "nav2_util/node_utils.hpp"
+
+namespace nav2_behaviors
+{
+AssistedTeleop::AssistedTeleop()
+: TimedBehavior<AssistedTeleopAction>(),
+  feedback_(std::make_shared<AssistedTeleopAction::Feedback>())
+{}
+
+void AssistedTeleop::onConfigure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // set up parameters
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "input_vel_topic", rclcpp::ParameterValue(std::string("input_vel_topic")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "joystick_topic", rclcpp::ParameterValue(std::string("joystick_topic")));
+
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("input_vel_topic", input_vel_topic_);
+  node->get_parameter("joystick_topic", joystick_topic_);
+}
+
+Status AssistedTeleop::onRun(const std::shared_ptr<const AssistedTeleopAction::Goal> command)
+{
+  command_time_allowance_ = command->time_allowance;
+  end_time_ = steady_clock_.now() + command_time_allowance_;
+
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::inputVelocityCallback,
+      this, std::placeholders::_1));
+
+  joy_sub_ = node->create_subscription<sensor_msgs::msg::Joy>(
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::joyCallback,
+      this, std::placeholders::_1));
+
+  return Status::SUCCEEDED;
+}
+
+void AssistedTeleop::onActionCompletion()
+{
+  RCLCPP_INFO(logger_, "Action completed");
+  input_twist_ = geometry_msgs::msg::Twist();
+  vel_sub_.reset();
+  joy_sub_.reset();
+}
+
+Status AssistedTeleop::onCycleUpdate()
+{
+  rclcpp::Duration time_remaining = end_time_ - steady_clock_.now();
+  if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
+    stopRobot();
+    RCLCPP_WARN(
+      logger_,
+      "Exceeded time allowance before reaching the BackUp goal - Exiting BackUp");
```

Comment:
Not backup :-)

Reply:
I think this is a success ending point though, no? Its done, its succeeded in teleopping for that period of time